js/eventloop.go

```diff
@@ -0,0 +1,165 @@
+// argument and wakeup the loop if needed.
+// If the eventLoop has since stopped it will not be execeted
+// This function *must* be called from within running on the event loop, but it's result can be called from anywhere
+// TODO better name
+func (e *eventLoop) reserve() func(func() error) {
+	e.lock.Lock()
+	e.reservedCount++
+	e.lock.Unlock()
+
+	return func(f func() error) {
+		e.lock.Lock()
+		e.queue = append(e.queue, f)
+		e.reservedCount--
+		e.lock.Unlock()
+		e.wakeup()
+	}
+}
+
+func (e *eventLoop) promiseRejectionTracker(p *goja.Promise, op goja.PromiseRejectionOperation) {
+	// No locking here as the goja runtime will call this synchrnously
+	// Read Notes on https://tc39.es/ecma262/#sec-host-promise-rejection-tracker
+	if op == goja.PromiseRejectionReject {
+		e.pendingPromiseRejections[p] = struct{}{}
+	} else { // goja.PromiseRejectionHandle so a promise that was previously rejected without handler now got one
+		delete(e.pendingPromiseRejections, p)
+	}
+}
+
+func (e *eventLoop) popAll() (queue []func() error, reserved bool) {
+	e.lock.Lock()
+	queue = e.queue
+	e.queue = make([]func() error, 0, len(queue))
+	reserved = e.reservedCount != 0
+	e.lock.Unlock()
+	return
+}
+
+// start will run the event loop until it's empty and there are no reserved spots
+// or a queued function returns an error. The provided function will be the first thing executed.
+// After start returns the event loop should not be reused if an error was returend
+func (e *eventLoop) start(f func() error) error {
+	e.queue = []func() error{f}
+	for {
+		queue, reserved := e.popAll()
+
+		if len(queue) == 0 {
+			if !reserved { // we have empty queue and nothing that reserved a spot
+				return nil
+			}
+			<-e.wakeupCh // wait until the reserved is done
+			continue
+		}
+
+		for _, f := range queue {
+			if err := f(); err != nil {
+				return err
+			}
+		}
+
+		// yes it will be at random but that seems to be the case in other tools as well
+		for promise := range e.pendingPromiseRejections {
+			// TODO maybe throw the whole promise up and get make a better message outside of the event loop
+			value := promise.Result()
+			// try to get the stack without actually needing the runtime
+			// this might break in the future :(
+			if o, ok := promise.Result().(*goja.Object); ok {
+				stack := o.Get("stack")
+				if stack != nil {
+					value = stack
+				}
+			}
+			// this is the de facto wording in both firefox and deno at least
+			return fmt.Errorf("Uncaught (in promise) %s", value) //nolint:stylecheck
+		}
+	}
+}
+
+// this exists so we can wait on all reserved so we know nothing is still doing work
+func (e *eventLoop) waitOnReserved() {
+	for {
+		_, reserved := e.popAll()
+		if !reserved {
+			return
+		}
+		<-e.wakeupCh // wait until the reserved is done
+	}
+}
+
+func (e *eventLoop) addSetTimeout(ctx context.Context, rt *goja.Runtime) {
```

Comment:
I might be biased by having worked on time mocking and testing in general, but I would like to see some tests for this specific method 😄

Reply:
All the tests for this are currently in `./core/local/` as it needs the who machinery in order to be reasonable to test it. I guess I can make small tests here as well, but I just don't think it makes much of a difference given that the other tests are much more encompassing.